File: src/frontend/src/api/gmail.ts
export interface GmailSearchRequest {
  query: string;
  max_results: number;
}

export interface GmailMessageSummary {
  id: string;
  thread_id: string;
  subject: string | null;
  from_email: string | null;
  date: string | null;
  snippet: string | null;
  label_ids: string[];
}

export interface GmailSearchResponse {
  messages: GmailMessageSummary[];
  result_size_estimate: number;
}

export interface GmailParsedBody {
  text: string;
  html: string | null;
}

export interface GmailMessageDetail {
  id: string;
  thread_id: string;
  subject: string | null;
  from_email: string | null;
  to: string[];
  cc: string[];
  date: string | null;
  snippet: string | null;
  body: GmailParsedBody;
  headers: Record<string, string>;
}

export interface GmailThreadDetail {
  id: string;
  messages: GmailMessageDetail[];
}

export interface EmailAddress {
  email: string;
  name: string | null;
}

export interface PrepareNewEmailRequest {
  thread_id: string | null;
  sender_email: string;
  to: EmailAddress[];
  cc: EmailAddress[];
  bcc: EmailAddress[];
  subject: string;
  body: string;
  signature_policy: string;
}

export interface PrepareReplyEmailRequest {
  thread_id: string | null;
  gmail_thread_id: string;
  reply_to_message_id: string;
  sender_email: string;
  to: EmailAddress[];
  cc: EmailAddress[];
  subject: string;
  body: string;
  signature_policy: string;
}

export interface EmailArtifactResponse {
  thread_id: string;
  work_item_id: string;
  artifact_id: string;
  version: number;
  maturity: string;
  content: {
    draft_type: "new_email" | "reply_email" | "forward_email";
    sender_email: string;
    to: EmailAddress[];
    cc: EmailAddress[];
    bcc: EmailAddress[];
    subject: string;
    body: string;
    signature_policy: string;
    gmail_thread_id: string | null;
    reply_to_message_id: string | null;
    source_message_id: string | null;
    additional_note: string | null;
  };
}

// 阶段 4 的 Gmail 客户端只负责前端到后端的类型化请求。
// OAuth token、Gmail access token 和 client secret 都留在后端，浏览器看不到。
export async function searchGmail(payload: GmailSearchRequest): Promise<GmailSearchResponse> {
  const response = await fetch("/api/gmail/search", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });

  if (!response.ok) {
    throw new Error(`Gmail search failed with ${response.status}`);
  }

  return response.json() as Promise<GmailSearchResponse>;
}

export async function readGmailMessage(messageId: string): Promise<GmailMessageDetail> {
  const response = await fetch(`/api/gmail/messages/${encodeURIComponent(messageId)}`);

  if (!response.ok) {
    throw new Error(`Gmail message fetch failed with ${response.status}`);
  }

  return response.json() as Promise<GmailMessageDetail>;
}

export async function readGmailThread(threadId: string): Promise<GmailThreadDetail> {
  const response = await fetch(`/api/gmail/threads/${encodeURIComponent(threadId)}`);

  if (!response.ok) {
    throw new Error(`Gmail thread fetch failed with ${response.status}`);
  }

  return response.json() as Promise<GmailThreadDetail>;
}

export async function prepareNewEmailDraft(
  payload: PrepareNewEmailRequest,
): Promise<EmailArtifactResponse> {
  const response = await fetch("/api/gmail/prepare/new", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });

  if (!response.ok) {
    throw new Error(`Prepare new email failed with ${response.status}`);
  }

  return response.json() as Promise<EmailArtifactResponse>;
}

export async function prepareReplyEmailDraft(
  payload: PrepareReplyEmailRequest,
): Promise<EmailArtifactResponse> {
  const response = await fetch("/api/gmail/prepare/reply", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });

  if (!response.ok) {
    throw new Error(`Prepare reply email failed with ${response.status}`);
  }

  return response.json() as Promise<EmailArtifactResponse>;
}

export async function deleteGmailMessage(messageId: string): Promise<void> {
  const response = await fetch(`/api/gmail/messages/${encodeURIComponent(messageId)}`, {
    method: "DELETE",
  });

  if (!response.ok) {
    throw new Error(`Delete email failed with ${response.status}`);
  }
}

File: src/backend/app/schemas/gmail.py
from typing import Any, Literal

from pydantic import BaseModel, Field


class GmailSearchRequest(BaseModel):
    """Gmail 搜索请求。"""

    query: str = Field(default="", max_length=500)
    max_results: int = Field(default=10, ge=1, le=25)


class GmailMessageSummary(BaseModel):
    """Gmail 搜索结果中的轻量邮件摘要（含发件人、主题、日期、摘要）。"""

    id: str
    thread_id: str
    subject: str | None = None
    from_email: str | None = None
    date: str | None = None
    snippet: str | None = None
    label_ids: list[str] = []


class GmailSearchResponse(BaseModel):
    """Gmail 搜索响应。"""

    messages: list[GmailMessageSummary]
    result_size_estimate: int = 0


class GmailParsedBody(BaseModel):
    """MIME 解析后的邮件正文。"""

    text: str
    html: str | None = None


class GmailMessageDetail(BaseModel):
    """Gmail 邮件详情。"""

    id: str
    thread_id: str
    subject: str | None = None
    from_email: str | None = None
    to: list[str] = []
    cc: list[str] = []
    date: str | None = None
    snippet: str | None = None
    body: GmailParsedBody
    headers: dict[str, str] = {}
    raw: dict[str, Any] = {}


class GmailThreadDetail(BaseModel):
    """Gmail 线程详情。"""

    id: str
    messages: list[GmailMessageDetail]


class EmailAddress(BaseModel):
    """邮件地址。"""

    email: str
    name: str | None = None


class EmailDraftPayload(BaseModel):
    """本地邮件 Artifact 的内容结构。"""

    draft_type: Literal["new_email", "reply_email", "forward_email"]
    sender_email: str
    to: list[EmailAddress] = []
    cc: list[EmailAddress] = []
    bcc: list[EmailAddress] = []
    subject: str
    body: str
    signature_policy: str
    gmail_thread_id: str | None = None
    reply_to_message_id: str | None = None
    source_message_id: str | None = None
    additional_note: str | None = None


class PrepareNewEmailRequest(BaseModel):
    """准备新邮件本地草稿的请求。"""

    thread_id: str | None = None
    sender_email: str
    to: list[EmailAddress]
    cc: list[EmailAddress] = []
    bcc: list[EmailAddress] = []
    subject: str
    body: str
    signature_policy: str


class PrepareReplyEmailRequest(BaseModel):
    """准备回复邮件本地草稿的请求。"""

    thread_id: str | None = None
    gmail_thread_id: str
    reply_to_message_id: str
    sender_email: str
    to: list[EmailAddress]
    cc: list[EmailAddress] = []
    subject: str
    body: str
    signature_policy: str


class PrepareForwardEmailRequest(BaseModel):
    """准备转发邮件本地草稿的请求。"""

    thread_id: str | None = None
    source_message_id: str
    sender_email: str
    forward_to: list[EmailAddress]
    subject: str
    additional_note: str | None = None
    signature_policy: str = "no_signature"


class EmailArtifactResponse(BaseModel):
    """本地邮件 Artifact 响应。"""

    thread_id: str
    work_item_id: str
    artifact_id: str
    version: int
    maturity: str
    content: EmailDraftPayload
